### PR TITLE
Cron cron cron

### PR DESCRIPTION
**Pull request by Rob-Johnson**
Adding support from cron-style schedule expressions:

- ScheduleStreams are now parameterized for each type of Schedule
- Added a 'Nextable' typeclass, with default members for Cron and ISO8601 schedules
- Added a CronParser, for parsing schedules defined in the regular time format
- you provide cron schedules to the same '/scheduler/iso8601' endpoint that you provide other jobs to. This is kind of nasty, and I wish the endpoint wasn't named like that, but I'd rather consumers had a single place to POST jobs to.

Still needs battle testing and tidying up in places, but works on my machine? I'll deploy it to mesosstage and post a couple of test jobs.

@tub could you have a look over it?

### CODE REVIEW

File: src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala [hunk 1]
```diff
@@ -1,9 +1,16 @@
 package org.apache.mesos.chronos.schedule
 
+import scala.util.{Try, Success, Failure}
 import org.joda.time.{DateTime, DateTimeZone, Period}
 import java.util.logging.Logger
 import org.joda.time.format.ISOPeriodFormat
+import org.joda.time.DateTime
 import java.util.TimeZone
+import java.time.ZonedDateTime
+import com.cronutils.parser.{CronParser => CronUtilsParser}
```
Comment (tub) [inline src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala:10]: Ooh, that's cool. Didn't know you could do that.

File: src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala [hunk 2]
```diff
@@ -33,8 +40,42 @@ object ISO8601Parser extends Parser{
...
+        ExecutionTime.forCron(parsed)
+      }
+    }
+
+    val dateForNextRun = cronExpression.map {
+      executionTime => executionTime.nextExecution(ZonedDateTime.now)
+    }.map {
+      //there is no timezone to take into account here; we just want the next run for the
+      //cron job.
+      zdt => new DateTime(zdt.toInstant().toEpochMilli(), DateTimeZone.forID("UTC"))
```
Comment (tub) [inline src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala:59]: what does zdt stand for?

Comment (tub) [reply src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala:59]: oh, ZonedDateTime...

File: src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala [hunk 3]
```diff
@@ -33,8 +40,42 @@ object ISO8601Parser extends Parser{
...
+    }
+  }
+}
+
 object ParserForSchedule {
+  val log = Logger.getLogger(getClass.getName)
   def apply(input: String): Option[Parser] = {
-    Some(ISO8601Parser)
+    if(ISO8601Parser.iso8601ExpressionRegex.pattern.matcher(input).matches()) {
+      log.info("ISO801 Parser!")
```
Comment (tub) [inline src/main/scala/org/apache/mesos/chronos/schedule/Parser.scala]: log.debug, also `ISO8601Parser.iso8601ExpressionRegex.pattern.matcher(input).matches()` is pretty fugly

**Review by tub — APPROVED** [review]
Looks good to me. Was going to suggest some more testing, but anything on top of what's here would be more like integration testing.

**Comment by Rob-Johnson** [comment]
@tub I've switched Dockerfile.deb_build to use Xenial as a base image so I can get Java 8 - any objections?

**Pull request merged**